### ISSUE TITLE
`useRouteParams` and `useRouteQuery` for Nuxt

### ISSUE THREAD

**Issue by manniL** [nuxt/nuxt#24788]
### Describe the feature

Especially when it comes to handling URL state, I've seen the wildest implementations.
It would be great to have two simple helper composables which create "reactive" router params and router query values.
That way, it would be even easier to sync URL state (and we could properly document + advise it).

This [already exists in VueUse](https://vueuse.org/router/useRouteParams/) but does not utilize `navigateTo`.

---

Related issue (was more bare-bone): https://github.com/nuxt/nuxt/issues/24365

### Additional information

- [X] Would you be willing to help implement this feature?
- [X] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by Atinux** [comment]
Linking https://github.com/nuxt/nuxt/pull/24369 since it has also been discussed there.

**Comment by manniL** [comment]
@Atinux Yes, exactly. The focus is to make it not only "reactive" but also writable, eventually syncing the URL ☺️ 

> I also think we can directly open a PR to vueUse to leverage navigateTo when used with the [@vueuse/nuxt](https://github.com/vueuse/vueuse/tree/main/packages/nuxt?rgh-link-date=2023-12-15T23%3A32%3A50Z)

That'd be a good idea too if possible! 
Though having these functionalities "more prominent" could also help improving DX. I know lots of people are using VueUse anyway - but I also know people are missing out on it 👀

**Comment by harlan-zw** [comment]
`useRouteHash` is also in VueUse, maybe worth implementing with these

**Comment by manniL** [comment]
@harlan-zw `useRouteHash` won't work during SSR though so I'd rather not and just refer to the vueuse composable in this case (mentioning the caveat) 🙊

**Comment by manniL** [comment]
To take a further look, I'll check if the VueUse composable work with SSR. If yes, we will document it. If not let's revisit.

Also cc'ing @posva here - do you think the composables (no matter if the VueUse ones or not) could be a nice "native router" solution too?

**Comment by manniL** [comment]
@posva Thanks a lot for your thorough and quick feedback 🔥 I'll write up creating the RFC on my list (but let's see *when the time allows* 🙈)

> Regarding `navigateTo()`: I don't know what are the implications of using router.push() while the app is running. I know that it can be used in some extra places and adds redirect codes but I imagine it has no impact when executed while the app is already running, which is when these composables (useRouteParams(), etc) would be used.

Yes, as long as the app is running on the client-side, that's all good (which is why e.g. the VueUse composables work already fine). The only downside is when using SSR and wanting to change the route, thus possibly requiring redirects (and similar).

**Comment by posva** [comment]
But when would you change the route through useParam() in SSR?

**Comment by manniL** [comment]
@posva Good question 🤔  I could imagine that as response of a successful API call. For example, a path like `/item/<PARAM>` which then does an API call and update the param to be the "correct one" stored in a DB?

**Comment by posva** [comment]
But in that case the application in already running, you don't router.push on the initial render

**Comment by manniL** [comment]
@posva I was thinking of a [scenario like this](https://stackblitz.com/edit/nuxt-issues-24788). When hitting it during SSR, a redirect would be needed. On client-side, the router should handle it

**Comment by posva** [comment]
I see, thanks for the code sample! That usage is not correct from router's perspective: it should happen in a navigation guard. It works on the client but it doesn't work on the server since a push ends up rendering a new page so it has to happen before mounting. I understand the information is coming from `useFetch()`, but it should be doable properly when using [Data Loaders](https://github.com/vuejs/rfcs/discussions/460)

**Comment by manniL** [comment]
@posva Would you suggest waiting for Data Loaders then? ☺️

**Comment by posva** [comment]
They are already awaited in a navigation guard

**Comment by P4sca1** [comment]
We should keep in mind that adding this feature to nuxt natively using the same auto-imported name could break existing applications.

**Comment by ArbazIrshad** [comment]
Any plan to implement this feature? because useRouteQuery is not working for me in Nuxt and is giving me 
`Invalid value used as weak map key`
